Add beatmap availability tracker component for multiplayer and playlists

frenzibyte opened the pull request:
Required for the upcoming multiplayer room user download states PR, as it relies on the tracker here to inform the server about the current beatmap availability state and display them, if not obvious.

This PR may be a bit too big to review, but I was very concerned about splitting it and cause confusion over what the split changes, as they may only make sense when they're bundled together with the main PR.

I have at least split each commit carefully and made them stand on their own, so reviewing by going through each commit may be desirable, and would hopefully make this easy to review.

Few things to note:
 - I've made the `MultiplayerBeatmapTracker` inherit `DownloadTrackingComposite` and have exposed a couple of members in that base class to use/override in the tracker for performing just like `ReadyButton`'s beatmap matching logic does, as I've realized `ReadyButton` is almost a copy of `DownloadTrackingComposite`.
 - As in https://github.com/ppy/osu/commit/8ebac53bba9b3e3a07dd65b261013ed2b66058f8, I've enforced all public members that hold a download progress value to become `double`, as the `float`/`double` confliction tripped me up while I was testing the component (setting progress to `0.4f` and asserting it is `0.4d`).
 - In testing tracker component, I was very unsure about whether to use "soleily" archive in `TestResources` and create beatmap info for it, or use "my love" beatmap in `TestBeatmap` and make an archive for it, but I've decided to do the latter as I wasn't comfortable with the changes done to do the former. A decision may need to be made at some point over whether to use "soleily" or "my love" for all tests, to avoid this happening again.
 - `DownloadTrackingComposite` in general does not get notified when a model gets imported in a non-downloading path (i.e. drag-drop), it only sets state to `LocallyAvailable` once the model is finished importing and is added to the database, i'll solve that in a separate PR as it is out of scope from here.

frenzibyte committed:
Make model manager in `DownloadTrackingComposite` protected

frenzibyte committed:
Add overridable method for verifying models in database

frenzibyte committed:
Add overridable method for checking local availability of current model

frenzibyte committed:
Add beatmap availability tracker component for multiplayer

frenzibyte committed:
Cache beatmap tracker and bind to selected item in `RoomSubScreen`

frenzibyte committed:
Make `ArchiveModelManager` import method overridable (for testing purposes)

frenzibyte committed:
Expose method for triggering filename-backed success in `APIDownloadRequest`

Exactly like in `APIRequest<T>`

frenzibyte committed:
Expose method for setting progress of archive download request

frenzibyte committed:
Enforce `double` type in the download progress path

Wasn't sure where to exactly put this, or whether to split it, but it's very small change to worry about, so I guess it's fine being here

frenzibyte committed:
Add beatmapset file containing the eaxct beatmap in `TestBeatmap` solely

frenzibyte committed:
Add test coverage for the new multiplayer beatmap tracker

bdach commented:
> I've decided to do the latter as I wasn't comfortable with the changes done to do the former.

you can't just say that without stating what the changes were. do you mind explaining this properly? what does the opaque unreviewable `.osz` do in this diff?

bdach commented on osu.Game.Tests/Online/TestSceneMultiplayerBeatmapTracker.cs:
```diff
@@ -0,0 +1,178 @@
+
+            public override async Task<BeatmapSetInfo> Import(BeatmapSetInfo item, ArchiveReader archive = null, CancellationToken cancellationToken = default)
+            {
+                while (!AllowImport.IsSet)
+                    await Task.Delay(10, cancellationToken);
```
why does this essentially spin in a loop, instead of properly waiting on the `ManualResetEvent` with a timeout? why use the event at this point if you're going to do that? might as well be a boolean flag?

a `TaskCompletionSource` or a `SemaphoreSlim` is the proper solution for this if you want an async wait.

frenzibyte replied:
yep, I didn't use `ManualResetEvent.Wait()` because it is not async and I was very unsure about whether it's fine using a synchronous one. `SemaphoreSlim` isn't exactly a "wait for set" type of class, but `TaskCompletionSource` looks to be exactly what I need!

bdach replied:
`SemaphoreSlim` is applicable. the analogue for setting would be to start with a semaphore with zero count and then call `Release()` when needed.

bdach commented on osu.Game/Online/API/APIDownloadRequest.cs:
```diff
@@ -28,13 +29,19 @@ protected override WebRequest CreateWebRequest()
 
-        protected APIDownloadRequest()
+        internal void TriggerSuccess(string filename)
         {
-            base.Success += onSuccess;
+            if (this.filename != null)
+                throw new InvalidOperationException("Attempted to trigger success more than once");
+
+            this.filename = filename;
+
+            TriggerSuccess();
         }
```
why does this exist in the base request, instead of near the test where this actually makes sense?

frenzibyte replied:
API requests generally expose an internal `TriggerSuccess` method for testing purposes (even `APIRequest<T>` where it allows to provide arbitrary content like in here as the result of the successful request), so why bother doing it in the test request if it should've been in the base request class itself anyways.

bdach replied:
disagree. the filename is an internal implementation detail of the request. fundamental difference between controlling the core content of the response for a test, and being able to control a temporary filename in a download request for a test.

frenzibyte replied:
fair point

bdach commented on osu.Game/Online/API/ArchiveDownloadRequest.cs:
```diff
@@ -10,15 +10,21 @@ public abstract class ArchiveDownloadRequest<TModel> : APIDownloadRequest
         public readonly TModel Model;
 
-        public float Progress;
+        public double Progress { get; private set; }
```
disagree. should have remained a `float`. there's no point having double precision for data as throwaway as this. now that we're on .NET 5 you could even make this a `half` and i wouldn't blink an eye.

frenzibyte replied:
oh, good reason. will revert then.

bdach replied:
...this is still `double`?

frenzibyte replied:
ah missed it (and it's coincidentally the one this conversation started at)

frenzibyte commented:
Yeah, my bad, the changes in the former consisted of making a method in `TestResources` for creating a beatmap info that matches a certain soleily beatmap, along with calculating the MD5 hash of that certain soleily beatmap in archive, which requires directly opening the archive and retrieving the data of that beatmap's `.osu` file.

So I decided to go with the latter which is pretty much easier as all I had to do was copy the beatmap content in `TestBeatmap` into a `.osu` file, and create an archive that contains it (which I've done very carefully with `SharpCompress` in the same way lazer exports archive files).

One downside for using the latter though is that if the content is changed in `TestBeatmap`, then the archive will become outdated, and `TestBeatmap` cannot open and read from that archive as it is in the game project, not in the tests project, free to hear your thoughts.

bdach commented on osu.Game/Online/DownloadTrackingComposite.cs:
```diff
@@ -46,25 +47,41 @@ private void load()
+        /// By default, this calls <see cref="IModelDownloader{TModel}.IsAvailableLocally"/>,
+        /// but can be overriden to add additional checks for verifying the model in database.
+        /// </summary>
+        protected virtual bool IsModelAvailableLocally() => Manager.IsAvailableLocally(Model.Value);
```
this was null-permissive before. now it'll crash if the `Manager` is null, which it can be, and it's annotated as such.

frenzibyte replied:
whoops, good point.

bdach commented on osu.Game/Online/DownloadTrackingComposite.cs:
```diff
@@ -127,30 +144,42 @@ private void attachDownload(ArchiveDownloadRequest<TModel> request)
+                    if (!item.Equals(Model.Value))
+                        return;
+
+                    if (!VerifyDatabasedModel(item))
+                    {
+                        State.Value = DownloadState.NotDownloaded;
+                        return;
+                    }
```
so this will silently fail and possibly keep importing copies of the same map with a bad checksum, with no indication to the user as to what's going wrong? sounds not ideal

frenzibyte replied:
that's how the multiplayer's download button behaves, but yeah I don't like the UX there either, logging error will do at least, not sure what the message should be other than a `the imported beatmap does not match online version`, any ideas?

bdach replied:
sounds tolerable for a start, certainly better than nothing

bdach commented on osu.Game/Online/Rooms/MultiplayerBeatmapTracker.cs:
```diff
@@ -0,0 +1,89 @@
+
+        public MultiplayerBeatmapTracker()
+        {
+            State.BindValueChanged(_ => updateAvailability());
+            Progress.BindValueChanged(_ => updateAvailability());
+            updateAvailability();
```
this is potentially unsafe if any transforms are added in value change callbacks of `availability`.

frenzibyte replied:
the code at fault would be the one binding to `availability` that early, it shouldn't do so until `LoadComplete`.

bdach replied:
value change callbacks propagate on the same thread on which the original thread that triggered the change happened, so i'm not sure how the consumer's load state is relevant?

frenzibyte replied:
I don't follow where the issue is, `State` and `Progress` are both changed on the update thread?

bdach commented on osu.Game/Online/Rooms/MultiplayerBeatmapTracker.cs:
```diff
@@ -0,0 +1,89 @@
+
+            BeatmapInfo matchingBeatmap;
+
+            if (databasedSet.Beatmaps == null)
```
when does this happen?

frenzibyte replied:
when deleting the beatmap, `ItemRemoved` would be invoked with a beatmapset that has `Beatmaps = null`.

bdach replied:
not sure how to feel about letting local bypasses like this spread but fair enough i guess

frenzibyte replied:
I thought the comment inside would be enough?

bdach replied:
it says what's happening, not why. i don't know what a "usable state" means. mention deleting in the comment, if the scenario this is fixing is about deleting.

frenzibyte replied:
fair point

frenzibyte replied:
actually, I've removed the whole thing as I realize now this method will only be called after importing.

tested with the ready button using the dependent PR #11516 and it's working without any crashes.

frenzibyte committed:
Use `TaskCompletionSource` for better awaiting

frenzibyte committed:
Enforce `float` type in the download progress path instead

frenzibyte committed:
Add null-permissive operator to manager back

frenzibyte committed:
Log important message if user imported bad-checksum beatmap

frenzibyte committed:
Make `TriggerSuccess(filename)` protected and expose in test instead

frenzibyte committed:
Merge branch 'master' into multiplayer-beatmap-tracker

frenzibyte committed:
Wait for scheduled state changes before asserting

bdach commented on osu.Game/Online/Rooms/MultiplayerBeatmapTracker.cs:
```diff
@@ -0,0 +1,99 @@
+        {
+            var verified = verifyDatabasedModel(databasedSet);
+            if (!verified)
+                Logger.Log("The imported beatmapset does not match the online version.", LoggingTarget.Runtime, LogLevel.Important);
```
```suggestion
                Logger.Log("The imported beatmap set does not match the online version.", LoggingTarget.Runtime, LogLevel.Important);
```

bdach commented:
i am still of the opinion that the `.osz` does not need to exist here and should be gone. there's an existing archive, if the only issue is calculating its md5, then i'm not sure it's an issue, unless i am shown code ugly enough to be convinced that that should not be done.

frenzibyte commented on osu.Game.Tests/Online/TestSceneMultiplayerBeatmapTracker.cs:
```diff
@@ -0,0 +1,181 @@
+
+        private void addAvailabilityCheckStep(string description, Func<BeatmapAvailability> expected)
+        {
+            // In DownloadTrackingComposite, state changes are scheduled one frame later, wait one step.
+            AddWaitStep("wait for potential change", 1);
+            AddAssert(description, () => tracker.Availability.Value.Equals(expected.Invoke()));
```
I intentionally didn't because really all that matters is waiting for one frame, which will always be achieved with adding a waiting step, as in the above comment.

bdach replied:
...what? you do realise a wait step is not one frame? surely an until step is more reliable?

frenzibyte replied:
perhaps I haven't explained properly, adding a wait step will delays the assert from being executed, to one frame next at most.

an until step will keep waiting for several frames if the tracker for some reason did not change availability in time and pass the test, even though that's not correct, so this way will be more accurate in terms of testing.

peppy replied:
This fails randomly still.

```
osu.Game.Tests.Online.TestSceneMultiplayerBeatmapAvailabilityTracker.TestBeatmapDownloadingFlow

TearDown : osu.Framework.Testing.Drawables.Steps.AssertButton+TracedException : state locally available 
--TearDown
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Graphics.Drawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Platform.GameHost.UpdateFrame()
   at osu.Framework.Platform.HeadlessGameHost.UpdateFrame()
   at osu.Framework.Threading.GameThread.ProcessFrame()
   at osu.Framework.Threading.GameThread.runWork()
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()




-----

One or more child tests had errors
  Exception doesn't have a stacktrace



34414733: Running MultiplayerBeatmapAvailabilityTracker (TestSceneMultiplayerBeatmapAvailabilityTracker) visual test cases...
34414733: TestSceneMultiplayerBeatmapAvailabilityTracker
34414933: step 1 Assert: ensure beatmap unavailable
34415133: step 2 Repeat: wait for potential change 1/1
34415333: step 3 Assert: state not downloaded
34415533: step 4 start downloading
34415733: step 5 Repeat: wait for potential change 1/1
34415933: step 6 Assert: state downloading 0%
34416133: step 7 set progress 40%
34416333: step 8 Repeat: wait for potential change 1/1
34416533: step 9 Assert: state downloading 40%
34416733: step 10 finish download
34416933: step 11 Repeat: wait for potential change 1/1
34417133: step 12 Assert: state importing
34417333: step 13 allow importing
34417533: step 14 Repeat: wait for import (1 tries)..............................................[database] 2021-02-01 08:39:47 [verbose]: [?????] Beginning import...
..............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................[database] 2021-02-01 08:39:48 [verbose]: [672a6] Validating online IDs for 12 beatmaps...
......................................................................................................................................................................................................................................................................................
[database] 2021-02-01 08:39:48 [verbose]: [672a6] Import successfully completed!
34616533: step 15 Repeat: wait for potential change 1/1
```

frenzibyte replied:
Right, I got overthrown by this near-impossible possibility that the `State` could be changed after the assert step is called because parent's scheduler runs before children. 

Turns out what the issue actually was is that the "wait for import" until step was relying on `IsAvailableLocally()`, so given that the import is performed in another thread, this method may return `true` early before the importing has finished and `ItemUpdated` is invoked. 

So I fixed it by retrieving the current import task and waiting for it until it completely finishes, and have made the test go through this "run selected tests until failure" feature for around an hour, with no sign of failure!

frenzibyte committed:
Remove unnecessary full querying

frenzibyte committed:
Enforce one missed property back to single-floating type

frenzibyte committed:
Rewerite beatmap term properly

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

frenzibyte commented:
Not sure what your problem is with just adding an archive that holds a single beatmap file with content copied exactly from `TestBeatmap`, but I'll try going soleily way (`TestResources`)

peppy commented on osu.Game/Online/DownloadTrackingComposite.cs:
```diff
@@ -46,25 +47,41 @@ private void load()
+        /// Verifies that the given databased model is in a correct state to be considered available.
+        /// </summary>
+        /// <example>
+        /// In the case of multiplayer/playlists, this has to verify that the databased beatmap set with the selected beatmap matches what's online.
```
This is broken (not sure how to read or what this function does).

frenzibyte replied:
Yeah as usual, still bad at documenting since day 1, what about this then:
```
In the case of multiplayer/playlists, this method will be overriden to verify that the imported beatmap set with the beatmap selected in room matches what's online.
```
or maybe remove the "matches what's online" and directly say "has the correct checksum set."

smoogipoo replied:
I'd suggest something like:
```csharp
/// <summary>
/// Checks that a database model matches the one expected to be downloaded.
/// </summary>
```

peppy committed:
Merge branch 'master' into multiplayer-beatmap-tracker

peppy commented on osu.Game/Online/Rooms/BeatmapAvailability.cs:
```diff
@@ -19,17 +19,17 @@ public class BeatmapAvailability : IEquatable<BeatmapAvailability>
 
         [JsonConstructor]
-        private BeatmapAvailability(DownloadState state, double? downloadProgress = null)
+        private BeatmapAvailability(DownloadState state, float? downloadProgress = null)
```
Should this not have been merged and part of the server-side deploy? If we were using anything other than json serialisation this is an API breaking change.

frenzibyte replied:
Right, was concerned about that. this is using JSON serialization, but for safety, I'll move this to a different PR and let server depend on it.

peppy replied:
It's not worth it at this point, fine to stay here. Just pointing it out.

peppy commented on osu.Game/Online/Rooms/MultiplayerBeatmapTracker.cs:
```diff
@@ -0,0 +1,89 @@
+            return verified;
+        }
+
+        private bool verifyDatabasedModel(BeatmapSetInfo databasedSet)
```
any reason this isn't just inline? feels really weird to split it out here.

peppy commented on osu.Game/Online/Rooms/MultiplayerBeatmapTracker.cs:
```diff
@@ -0,0 +1,89 @@
+
+namespace osu.Game.Online.Rooms
+{
+    public class MultiplayerBeatmapTracker : DownloadTrackingComposite<BeatmapSetInfo, BeatmapManager>
```
xmldoc explaining what this class intends to do

frenzibyte committed:
Inline with above method

frenzibyte committed:
Add xmldoc explaining what the multiplayer beatmap tracker is for

frenzibyte committed:
Fix wrong importing of test beatmaps

Importing via `testBeatmapSet` causes the beatmapset hash to not be calculated due to no files existing in the importing process, which leads into not reusing existing test beatmaps due to no hash.

frenzibyte committed:
Use beatmap "soleily" and remove no longer needed archive

peppy committed:
Merge branch 'master' into multiplayer-beatmap-tracker

peppy committed:
Add signalr messagepack key attribute

peppy committed:
Rename class to better match purpose

peppy commented on osu.Game/Screens/OnlinePlay/Match/RoomSubScreen.cs:
```diff
@@ -40,6 +40,17 @@ public abstract class RoomSubScreen : OnlinePlaySubScreen, IPreviewTrackOwner
+
+        protected RoomSubScreen()
+        {
+            InternalChild = BeatmapAvailablilityTracker = new MultiplayerBeatmapAvailablilityTracker
```
I really don't like this. Feels like it could go wrong at any point (overwriting other children etc). Use `AddInternal` instead. And probably move to BDL below.

frenzibyte replied:
right, was thinking this will be the top constructor, but that's actually bad practice regardless

frenzibyte committed:
Fix "wait for import" until step potentially finishing early

If not obvious, the issue with previous code is that it was checking for `IsAvailableLocally`, while the import is happening on a different thread, so that method could return `true` before the importing has finished and `ItemUpdated` event is called.

frenzibyte committed:
Move tracker loading into BDL

frenzibyte committed:
Fix using private constructor on MessagePack object

frenzibyte committed:
Apply documentation suggestion

smoogipoo commented on osu.Game/Online/DownloadTrackingComposite.cs:
```diff
@@ -46,25 +47,41 @@ private void load()
+        /// Checks that a database model matches the one expected to be downloaded.
+        /// </summary>
+        /// <example>
+        /// In the case of multiplayer/playlists, this has to check that the databased beatmap set with the selected beatmap matches what's online.
```
```suggestion
        /// For online play, this could be used to check that the databased model matches the online beatmap.
```

smoogipoo reviewed (changes requested):
This looks fine to me otherwise

smoogipoo commented:
This looks good from me otherwise :)

frenzibyte committed:
Enhance documentation

Co-authored-by: Dan Balasescu <smoogipoo@smgi.me>

smoogipoo committed:
Rename to OnlinePlayBeatmapAvailabilityTracker

smoogipoo committed:
Merge branch 'master' into multiplayer-beatmap-tracker
